connman/vpn/vpn-provider.c

```diff
@@ -3289,6 +3308,36 @@ const char *vpn_provider_get_path(struct vpn_provider *provider)
 unsigned int vpn_provider_get_authentication_errors(
 						struct vpn_provider *provider)
 {
+	const char *val;
+	unsigned int limit;
+	gint64 diff;
+
+	val = vpn_provider_get_string(provider, "AuthErrorLimit");
+	if (val)
+		limit = g_ascii_strtoull(val, NULL, 10);
+	else
+		limit = ALLOW_AUTH_ERRORS_DEFAULT;
+
+	diff = g_get_monotonic_time() - provider->previous_connect_time;
```

Comment:
Other option is to use the system time, which can be changed when it is updated via NTP.

Yeah, this may not be really optimal solution yet. Maybe the value just should be checked for being below the `CONNECT_OK_DIFF` to trigger the check as well.

Reply:
Added a check for using the `auth_error_counter` when the monotonic time is less than the `CONNECT_OK_DIFF` with no previous connection time set: 45e1bdeace1e1ebd61090c78d30947ab0d48d635 - I guess this would suffice for that case you brought up?

Reply:
I guess it might suffice, but ...

It depends on what that glib thingy really does. If it uses CLOCK_MONOTONIC, it does among other things not advance when device is suspended. Which - come to think about it - might actually be bigger problem as the 1h limit can get arbitrary long.

One way to normalize situation could be to: replace that glib function with something that uses CLOCK_BOOTTIME and init / reset that previous_connect_time to -CONNECT_OK_DIFF.

Reply:
Oh really, that is stopped while in suspend? Great.

I just actually noticed that in some cases OpenVPN may throw a improper shutdown error -> connection error -> reset the connection success time back to 0 and when I get to trigger the AUTH_FAILED reported back by a VPN provider it again clears credentials for no reason. This becomes quite tricky,

Reply:
So there can be two scenarios:

1) server does not get the proper shutdown cmd from client -> next auths for X period of time are replied with AUTH_FAILED (this scenario is handled now)

2) client binary dies in improper way when disconnecting and after this the server detects the client not being valid for the next auths -> the connection error reset last successful connection timer -> auths are cleared without any valid reason when server again replies with AUTH_FAILED.

I'm starting to think that this AUTH_FAILED is indeed related to servers that allow only one active VPN from an account. And the timeouts, especially with UDP, make this a lengthy process to avoid false positives in active connection detection.  I'm assuming the servers clean the client info with clean shutdown. But that other issue, scenario 2, is because of how the plugin + binary work together.

Reply:
Thinking out loud here: but then again it is completely wrong to accept connection errors when the VPN is already disconnected and residing in IDLE state as it was the case in 2). That seems to happen when WiFi 1 is connected and then rapidly one changes to WiFi 2.

Reply:
I'd assume a365748da10f061b82ca6fd52d9e41f870600d82 prevents the 2) from happening now. Getting the problems reproduced takes a long time and effort..

Reply:
Seemed to work as I finally got the issue to reproduce. But yeah, the ambiguous statement on `g_get_monotonic_time()` about sleep is indeed true. Now getting the uptime from `/proc/uptime` would be an option but.. is it really feasible. In case of a sleeping device it might be that 1h becomes more than 2h with the current approach.

Reply:
@spiiroin Replaced the monotonic clock with call to sysinfo() 5edf3dfd999be2975fe1fb35153ff7ca6d9433a9 - so now it deals in uptime. That should work now well.

Reply:
> sysinfo() ... That should work

probably does, just wondering why you chose such a big gun (gets all kinds of things in addition to uptime) instead of clock_gettime(CLOCK_BOOTTIME)?

Reply:
Because Harry Callahan also liked to wear a big gun!

Yeah, that `clock_gettime()` is even better, neater.

Reply:
Changed and updated the whole thing with documentation and all. Any more concerns/comments?

Reply:
> Any more concerns/comments?

nope